Closes #3207: Add new variable 'mysql-tls_version' controlling the 'tls versions' allowed for client connections

JavierJF opened the pull request:
Closes #3207.

JavierJF committed:
Closes #3207: Add new variable 'admin-tls_version' controlling the 'tls versions' allowed for client connections

JavierJF committed:
Added new test 'test_admin-tls_version-t' for new introduced variable 'admin-tls_version-t'

renecannao commented on lib/ProxySQL_Admin.cpp:
```diff
@@ -7029,6 +7031,7 @@ char * ProxySQL_Admin::get_variable(char *name) {
 		sprintf(intbuf, "%d", variables.p_memory_metrics_interval);
 		return strdup(intbuf);
 	}
+	if (!strcasecmp(name,"tls_version")) return s_strdup(variables.tls_version);
```
Why `s_strdup` and not `strdup` ?

JavierJF replied:
It's used in `ProxySQL_Admin::get_variable` for the same purpose in multiple places:

- https://github.com/sysown/proxysql/blob/v2.1.1/lib/ProxySQL_Admin.cpp#L7035
- https://github.com/sysown/proxysql/blob/v2.1.1/lib/ProxySQL_Admin.cpp#L7036
- https://github.com/sysown/proxysql/blob/v2.1.1/lib/ProxySQL_Admin.cpp#L7037
- https://github.com/sysown/proxysql/blob/v2.1.1/lib/ProxySQL_Admin.cpp#L7040
- https://github.com/sysown/proxysql/blob/v2.1.1/lib/ProxySQL_Admin.cpp#L7066

Probably because it performs and extra check on the passed pointer to check for `nullity` before calling the actual `strdup`.

renecannao commented on lib/ProxySQL_Admin.cpp:
```diff
@@ -548,6 +548,7 @@ static char * admin_variables_names[]= {
 	(char *)"checksum_admin_variables",
 	(char *)"restapi_enabled",
 	(char *)"restapi_port",
+	(char *)"tls_version",
```
This variable is only used by the MySQL worker threads.
It must be a `mysql-` variable, not an `admin-` variable

renecannao commented on include/proxysql_admin.h:
```diff
@@ -175,6 +181,7 @@ class ProxySQL_Admin {
 		int web_port;
 		int web_port_old;
 		int p_memory_metrics_interval;
+		char* tls_version;
```
This variable is only used by the MySQL worker threads.
It must be a `mysql-` variable, not an `admin-` variable

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -4731,6 +4731,39 @@ void MySQL_Session::handler___status_CHANGING_USER_CLIENT___STATE_CLIENT_HANDSHA
+ * @return In case of success, the OpenSSL identifier for the supplied
+ *   'TLS' version, '-1' otherwise.
+ */
+int string_to_tls_version(const std::string& tls_version) {
```
tls version is not case sensitive.
This function will fails if a different case is used.

JavierJF replied:
This was intended at the time, but I checked and looks like the original behavior it's indeed case insensitive, changing it.

renecannao commented on lib/ProxySQL_Admin.cpp:
```diff
@@ -7084,6 +7087,46 @@ void ProxySQL_Admin::delete_credentials(char *credentials) {
+ * @param tls_version_val The value of 'tls_version' variable to be verified.
+ * @return 'true' if the value was found to be correct, false otherwise.
+ */
+bool valid_tls_versions_var(const std::string& tls_version_val) {
```
tls version is not case sensitive.
This function will fails if a different case is used.

JavierJF replied:
As in comment https://github.com/sysown/proxysql/pull/3359#discussion_r601211793.

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -4755,6 +4788,17 @@ void MySQL_Session::handler___status_CONNECTING_CLIENT___STATE_SERVER_HANDSHAKE(
 		client_myds->wbio_ssl = BIO_new(BIO_s_mem());
 		client_myds->ssl=SSL_new(GloVars.global.ssl_ctx);
+		// get the current 'tls_version' specified in the global variable 'admin-tls_version'
+		std::string tls_version_var = GloAdmin->get_variable(const_cast<char*>("tls_version"));
```
A MySQL worker thread (in `MySQL_Thread`, `MySQL_Session`, or otherwise), should never call `GloAdmin->get_variable()` without getting a global lock on `GloAdmin` itself, otherwise dirty reads can happen.
In fact, a MySQL worker thread should never access admin variables.
This should be implemented as a `mysql-` , `mysql-tls_version`, and follow the same pattern of other mysql variable: a global variable in `MySQL_Threads_Handler`, and each thread should have its own local storage copy in `mysql_thread___tls_version` that can access without locking.

JavierJF replied:
I thought about this, but since the original request specifically asked for an Admin variable I went that way. I also thought about the dirty reads, but since this is something that happens per connection, didn't wanted to introduce a lock there. Thanks for the feedback, changing it in favor of 'mysql-tls_version'.

renecannao commented on test/tap/tests/test_admin-tls_version-t.cpp:
```diff
@@ -0,0 +1,292 @@
+/**
+ * @brief Copy of the supported 'tls_versions' by ProxySQL.
+ */
+const std::array<const char*, 4> valid_tls_versions {
```
valid versions do not include any value with different cases, like `tlsv1.3` ot `tLsV1.1`

JavierJF committed:
Removed 'admin-tls_version' variable in favor of not-case sensitive 'mysql-tls_version'

JavierJF committed:
Reshaped 'test_admin-tls_version-t.cpp' for 'admin-tls_version' into 'test_mysql-tls_version-t.cpp' for 'mysql-tls_version'

JavierJF commented:
Retest this please

JavierJF committed:
Updated 'test_mysql-tls_version-t.cpp' file documentation

JavierJF committed:
Merge branch 'v2.1.1' into v2.1.1-3207

JavierJF committed:
Merge branch 'v2.x' of github.com:sysown/proxysql into v2.1.1-3207

JavierJF commented:
retest this please

JavierJF committed:
Merge branch 'v2.x' of github.com:sysown/proxysql into v2.1.1-3207

JavierJF committed:
Multiple changes to 'mysql-tls_version' implementation #3207

+ Changed default allowed versions to 'TLSv1.2' and 'TLSv1.3'. This
  change is required due to the recent upgrade to OpenSSL 3.0, which
  deprecated older protocol versions.
+ Added capability for setting older protocol versions than 'TLSv1.2'
  issuing a warning to the user.
+ Refactored extraction of allowed versions from 'mysql_thread___tls_version'
  variable and other minor improvements.

JavierJF committed:
Make 'test_mysql-tls_version-t' compatible with OpenSSL 3.0 #3207

+ Test is now compiled against 'libmysqlclient' due to 'libmariadbclient'
  dependency being also compiled against 'OpenSSL 3.0'.
+ Minor refactor and reuse of function from utils.

JavierJF commented:
Retest this please.

JavierJF committed:
Add conditional compilation to 'test_myqsl-tls_version'

- Added conditional compilation for 'test_mysql-tls_version' left as an
  example on how to compile tap tests against libmysqlclient.
- Homogenized compilation line in Makefile

pondix commented:
Automated message: PR pending admin approval for build testing

mirostauder commented:
retest this please

mirostauder commented:
retest this please

linuxmail commented:
hi,

is there some activity on it ? :-) It would be very nice to have this options.

cu denny

mirostauder commented:
Can one of the admins verify this patch?
